lib/ret/media_resolver.ex

```diff
@@ -45,6 +48,40 @@ defmodule Ret.MediaResolver do
     resolve_non_video(query, root_host)
   end
 
+  # For youtube.com, we need to rate limit requests. Only do one at a time on this host.
+  # Also compute ttl here based upon expire, to localize youtube.com specific logic.
+  def resolve(%MediaResolverQuery{} = query, "youtube.com" = root_host) do
+    Mutex.under(MediaResolverMutex, :youtube, @youtube_rate_limit_timeout, fn ->
+      now = System.system_time(:millisecond)
+      res = resolve_with_ytdl(query, root_host, query |> ytdl_format(root_host))
+
+      {:commit, %Ret.ResolvedMedia{uri: %URI{query: youtube_query}} = resolved_media} = res
+
+      # YouTube returns a 'expire' which has timestamp of expiration.
+      resolved_media =
+        with parsed_youtube_query <- URI.decode_query(youtube_query),
+             expire when is_binary(expire) <- Map.get(parsed_youtube_query, "expire"),
+             {expire_s, _} <- Integer.parse(expire),
+             # Don't use custom TTL if expires in < 2 minutes
+             ttl_s when ttl_s > 120 <- expire_s - System.system_time(:second) do
```

Comment:
Did you actually see this? In the videos I tested expire always seemed to be 6 hours. In any case, probably makes sense to use the ttl regardless no? Since the link will literally be broken after it expires, so caching it is wrong.

Reply:
thats a good point, i may have had the wrong mental model. i thought that the URL returned by youtube-dl would be the same up until expiration, so eventually the window would tighten. beyond that tho the other point you raise is right, i'm not totally sure what i thought i was accomplishing here- probably in part due to paternity brain

Reply:
No I think youtube-dl will give you a new url each time, just the way youtube's CDN and such work is that the signed url it generates is only valid for a certain window (seemingly always 6 hours right now).